jetty-core/jetty-http/src/main/java/org/eclipse/jetty/http/CachingHttpContentFactory.java

```diff
@@ -0,0 +1,586 @@
+            {
+                wasAdded.set(true);
+                CachingHttpContent cachingContent = (httpContent == null) ? newNotFoundContent(p) : newCachedContent(p, httpContent);
+                _cachedSize.addAndGet(cachingContent.getBytesOccupied());
+                return cachingContent;
+            });
+
+            cachingHttpContent.retain();
+            if (wasAdded.get())
+                shrinkCache();
+            return (cachingHttpContent instanceof NotFoundHttpContent) ? null : cachingHttpContent;
+        }
+        return httpContent;
+    }
+
+    protected CachingHttpContent newCachedContent(String p, HttpContent httpContent)
+    {
+        return new CachedHttpContent(p, httpContent);
+    }
+
+    protected CachingHttpContent newNotFoundContent(String p)
+    {
+        return new NotFoundHttpContent(p);
+    }
+
+    protected interface CachingHttpContent extends HttpContent
+    {
+        long getLastAccessedNanos();
+
+        void setLastAccessedNanos(long nanosTime);
+
+        String getKey();
+
+        boolean isValid();
+
+        void retain();
+    }
+
+    protected class CachedHttpContent extends HttpContentWrapper implements CachingHttpContent
+    {
+        private final ByteBuffer _buffer;
+        private final String _cacheKey;
+        private final HttpField _etagField;
+        private final long _contentLengthValue;
+        private volatile long _lastAccessed;
+        private final Set<CompressedContentFormat> _compressedFormats;
+        private final String _lastModifiedValue;
+        private final String _characterEncoding;
+        private final MimeTypes.Type _mimeType;
+        private final HttpField _contentLength;
+        private final Instant _lastModifiedInstant;
+        private final HttpField _lastModified;
+        private final Retainable.ReferenceCounter _referenceCount = new Retainable.ReferenceCounter();
+
+        public CachedHttpContent(String key, HttpContent httpContent)
+        {
+            super(httpContent);
+            _cacheKey = key;
+
+            // TODO: do all the following lazily and asynchronously.
+            HttpField etagField = httpContent.getETag();
+            String eTagValue = httpContent.getETagValue();
+            if (StringUtil.isNotBlank(eTagValue))
+            {
+                etagField = new PreEncodedHttpField(HttpHeader.ETAG, eTagValue);
+            }
+            _etagField = etagField;
+
+            // Read the content into memory if the HttpContent does not already have a buffer.
+            ByteBuffer byteBuffer = httpContent.getByteBuffer();
+            if (byteBuffer == null)
+            {
+                try
+                {
+                    long contentLengthValue = getContentLengthValue();
+                    if (contentLengthValue > _maxCachedFileSize)
+                    {
+                        byteBuffer = _byteBufferPool.acquire((int)contentLengthValue, false);
```

Comment:
This cast to `int` should be handled better, as we don't want to blindly cast, but check if `contentLengthValue < Integer.MAX_VALUE` and only then cache it.

Reply:
I think this check should instead be `(contentLengthValue <= _maxCachedFileSize)`, and `_maxCachedFileSize` is an `int` so it will be <= `Integer.MAX_VALUE`.

But we also do this check in `isCacheable` so there should be no way to get here with a contentLengthValue greater than `Integer.MAX_VALUE`.

Reply:
So you're saying if the `maxCachedFileSize > Integer.MAX_VALUE` the entry should be added to the cache but there should be no buffer in the cache? So we will be caching `lastModified` `contentLength` etc.. but not caching the content.